src/Microsoft.AspNetCore.Mvc.TagHelpers/Cache/DistributedCacheTagHelperService.cs

```diff
@@ -74,20 +77,62 @@ HtmlEncoder HtmlEncoder
 
                             value = await _formatter.SerializeAsync(formattingContext);
 
-                            await _storage.SetAsync(key, value, options);
+                            
+                            using (var buffer = new MemoryStream())
+                            {
+                                // The stored content is 
+                                // - Length of the serialized cache key in bytes
+                                // - Cache Key
+                                // - Content
+
+                                var keyLength = BitConverter.GetBytes(serializedKey.Length);
+
+                                await buffer.WriteAsync(keyLength, 0, keyLength.Length);
+                                await buffer.WriteAsync(serializedKey, 0, serializedKey.Length);
+                                await buffer.WriteAsync(value, 0, value.Length);
+
+                                await _storage.SetAsync(storageKey, buffer.ToArray(), options);
```

Comment:
Instead of `ToArray` just use `TryGetBuffer`. It'll always return true since the memory stream is writable.